<a href="https://colab.research.google.com/github/aahFancyGun/dataScienceClass/blob/master/Copy_of_Unit1NotesSF.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Unit 1: Python Crash Course

This is a crash course in the basics of python and data manipulation.

Our goal is get vulnerable and exposed to some common ways to manipulate data in python, and learn some practical skills along the way.

We will:

1. Use variables to do basic operations.
2. Explore the built in data structures we'll use: int, float, string, list, and dict
3. Create and Manipulate Numpy Arrays
4. Import and Manipulate Pandas Dataframes (plus cover bools)
5. Use for loops to support Dataframe queries
6. Export Data from Colab

Will this unit cover all the python you'll need to ace this course?

No.

Will this feel like way too much information?

Probably.

But don't worry, you've got this! We'll learn as we go, and as we go you'll see me consisitently type "how to \___ in python" or "how to \___ with pandas" into google. You will learn to do the same.

Let this be Lesson 0: **you don't need to know everything, you just need to know enough to ask for what you need**.


##Arithmetic with ints
Let's start your journey into the world of data science with good ole' fashion addition, subtraction, and mutliplication. Let yourself wonder why I'm saving division for later.

In [ ]:
#Adding in python works inuitively. Not that python automatically prints the result of any operation
3 + 2

In [ ]:
#same with subtraction
3 - 2

In [ ]:
# multiplication use the astrisk symbol, "*"
3 * 2

In [ ]:
# exponetiation is quick too--just use the multiplication symbol twice
3 ** 2

This next code block introduces our first built in function, "type". Functions take in an input through parentheses, and then return an output.

In [ ]:
type(3)

int

For example, the absolute value function is "abs" in python, and works exactly like you think it would. Use the dash key "-" in front of a number (int or float) to make it negative

In [ ]:
abs(- 2)

I can stores these numbers in a variable for later use.

In [ ]:
# I store "-5" in the variable x
x = -5
# I store "2" in the variable thingamabob
thing = 2

#I print the content of the variables x and thing
x, thing

We can update the content of these variables whenever we want.

In [ ]:
# add "-5" to "2" using the variables x and thing
x = x + thing

#print the contents of x
x

Notice that we managed to get our computer to calculate $-5 +2$ without having to write $-5$ or $2$.

What happens if we run the same code block?

In [ ]:
# add the contents x and thingamabob
x = x + thing

#print the contents of x
x

**Task1**:

Explain why the variable x's value is now -1.

**Answer1**:

The value of x is defined and modified in a previous code block. In the previous code block it is initialized to the value of -5 and another variable is initalized to the value of 2. In the following code block x is modified by adding thing to it, meaning that the value will then be -3. This happens once more resulting in x being equal to -1.

Computers are only ever capable of doing exactly what we tell them to do, and its always possible to tell them to do something we don't underestand. This can be a

- feature: a good thing we may or may not have intended to happen.

or a

- bug: a bad thing we didn't want to happen, but are never the less entirely responsible for.

Every piece of code you'll ever write is either a feature, a bug, or confusing. Keep this in mind as we move through the course.

### Division

Let's look at an example of division, the other mathematical operation I left out.

In [ ]:
#use a backslash for division
-4 / 2

That number has a decimal; let's check to see if it's still an int.

In [ ]:
type(-4 / 2)

### Libraries

To better deal with this new float thing, I'll call a library of functions called numpy, and nickname it np for short.

I need to use numpy for two reasons:

- floats, as we'll see, do not behave as nicely as ints, and need to be treated with care. We'll only ever use libraries that show our floats the proper care and attention they need
- python doesn't really have any built in math functions; abs() is basically the only one.

In [ ]:
import numpy as np

With that our of the way, let's learn why we need to be cautious with floats.

##Floats, and the perils of infinite math with finite computers

We often represent the real numbers, like $\pi$ and $\sqrt{2}$, using another data type called a float. That's short for floating point number.

In [ ]:
#numpy has pi to a considerable number of digits stored, and ready to use
np.pi

In [ ]:
#it has a decimal, which in python means its a float
type(np.pi)

In the abstract world of math, $\pi$ has a precise, if infinitely long, answer.

Infinity is problematic for our finite computers.

3.141592653589793 is an accurate approximation, but if we multiply $\pi$  by itself enough--say, 31 times--then we can run into trouble.

I'm going to show you three different operations, all of which are mathematically equivalent to $\pi^{31}$.

Don't worry about understanding what the code blocks are doing. Just choose to believe me when I say that all of them should, theoretically, output the same thing.





First, regular exponetiation.

In [ ]:
import numpy as np

# I multipy pi by itself 31 times. In other words, I take pi to the 31st power
pi_1 = np.pi**(31)
pi_1

2580156526864955.5

Next, some expontial and logarithm shenanigans to turn exponetiaon into multiplication

Note: $e^{x}=\exp{(x)}$

In [ ]:
# this is mathematically equivalent to pi times itself 31 times
pi_2 = np.exp(31  * np.log(np.pi))
pi_2

2580156526864956.0

More exponential and logarithmic shenanigans, this time so that we aren't even multiplying, just adding.



In [ ]:
# this is mathematically equivalent to pi times itself 31 times
pi_3 = np.exp(np.exp(np.log(31) + np.log(np.log(np.pi))))
pi_3

2580156526864974.5

**Task2**:

Because of all the extra complex operations the program is more likely to truncate resulting in a loss of information.

**Answer2**:



##Strings, Lists, and Dicts
Before we get more into those libraires like numpy, let's return to base python.
Everything in this section can be done without importing a library.

###Strings
Let's start with strings. We can define a string by putting ' ' or " " around any character, including numbers

In [ ]:
'5', type('5'), 5, type(5)

In [ ]:
'type(5)', type('type(5)')

We'll use strings to define labels for our models and graphs, and to represent non-numerical data.

For right now though, they are our first example of an iterable data type. In other words, python keeps track of the individual characters in string, their order, and how many characters there are.

In [ ]:
"dog"

In [ ]:
#find the length of an iterable object
len('dog')

In [ ]:
#grab the first character in the string 'cat'
'dog'[0]

Variables are not strings. We'll get an error if we make that mistake

In [ ]:
dog

That said, I can assign a string to a variable.

In [ ]:
#assign the string 'dog' to variable x
x = 'dog'
#grab the second and third characters of the string
x[1:3]

Using the [ ] directly behind an iterable object, or directly behind a varable that points to an iterable object, splices the object according to some built in rules. Let's practice to see how it works.

**Task3**:

Splice the string 'Hippopotomonstrosesquippedaliophobia' to output 'monstro' using a single line of code.

Note: with data science, its not important to get it right on the first try. Just try stuff and see what happens.

In [ ]:
#ANSWER3

hippo = 'Hippopotomonstrosesquippedaliophobia'
hippo = hippo[9:16]
print(hippo)

monstro


There's one more thing about strings you must know: strings are immutable, unchangeable. Anytime it looks like you're changing a string, you're actaully creating a new string based off the old, still extant, string.

This means that we, as data scientists, will rarely, if ever, store our data in strings. They're also really hard to look at when they get larger, as Task3 showed.

###Lists

This brings us to our next data type, lists. They are iterable like strings, but they are mutable (changable), and can hold all types of data, not just characters.

In [2]:
x = 2
#define my list with brackets
my_list = ["dog", x, 5]
#print my list
my_list

['dog', 2, 5]

Notice that when I put a variable in my list, the list reports the contents of the variable. Let's see how that works

In [3]:
b_list = ['Hippopotomonstrosesquippedaliophobia'[9:16], 4/2 , len(my_list), my_list]

b_list

['monstro', 2.0, 3, ['dog', 2, 5]]

In [5]:
pi_3 = 1
#add another item to the end of my list
my_list.append(pi_3)

In [6]:
b_list[3][3]

1

**Task4**:

Explain what happened in the last three code blocks; why is 2580156526864974.5 in b_list now?

**Answer4**:

pi_3, which is equivalent to the number mentioned above was appended to the list via the function my_list.append(pi_3). But, because my_list is itself a part of b_list, it is also able to be accessed via b_list[3][3].

Let's practice splicing once more before moving on.

**Task5**:

Using one line of code, grab the last three entries of my_list without writing my_list anywhere in your code.

This task is a little harder than the others, so make sure you're working together. Don't be shy about googling 'how to splice lists in python' either.

In [8]:
#ANSWER5

newlist = [b_list[3][1], b_list[3][2], b_list[3][3]]
newlist

[2, 5, 1]

###Dicts

Like lists, dicts can hold any and all types of data. What differentiaties them from lists is that data in a dict is stored as a key/value pair called an item.

In [ ]:
#make my dict
my_dict = {'one':4/2, 2:my_list, '3':b_list[0]}
my_dict

Dicts aren't entirely iterable.

In [ ]:
my_dict[0]

Instead of splicing, we acces the values of a dict using its keys.

In [ ]:
my_dict['one'], my_dict['3'], my_dict[2.]

In [ ]:
#dicts can recognize when floats are equivalent to ints
my_dict[2]

If we want to iterate a dict, we need its keys. We can also just grab the values directly

In [ ]:
my_dict.keys()

In [ ]:
my_dict.values()

There are two other basic iterable data types in python: sets and tuples. We don't need to know much about them for this course though, so we'll move on to some data structures that we will use *a lot*.

im## Numpy Arrays

Nearly all scientific computing and data science in python uses numpy arrays, or something built out of numpy arrays. Numpy arrays are similar to lists, except they are imutable, and only hold one data type at a time

In [16]:
import numpy as np
x = 2

first_array = np.array([x, 2, 2, 3, 'dog', 'montro', 123.121])
type(first_array[0])

numpy.str_

In [15]:
hoo_array = np.array([np.pi, 11, 2 - 7])
type(hoo_array[1])

numpy.float64

hoo_array converted the two ints, 11 and -5, into floats, to ensure that everything is the same data type.

A similar thing happens when you try to mix floats and ints with strings.

**Task6**:

What happens when you make an numpy array with floats, strings, and ints?

In [31]:
#ANSWER6

#It will convert all elements of the array to a single variable type.

import numpy as np
x = 2

first_array = np.array([x, 2, 2, 3, 'dog', 'montro', 123.121])

for i in range(first_array.size):
  print(type(first_array[i]))

<class 'numpy.str_'>
<class 'numpy.str_'>
<class 'numpy.str_'>
<class 'numpy.str_'>
<class 'numpy.str_'>
<class 'numpy.str_'>
<class 'numpy.str_'>


**Answer6 cont.**:

[*write your answer here*]

Why make such a big deal out of numpy arrays, if they have this weird behavior?

Two reasons:
- ensuring that arrays are always of the same data type makes operations with arrays fast, which is essential when dealing with even moderately sized data sets.
- The automatic syncing of data formats has other benefits; see below

In [ ]:
#the smaller numbers are written in scitific notation, just like the larger one
#This encourages us to think about the relative scale of our values
tood_array = np.array([pi_3, 11, 2 - 7])
tood_array

array([ 2.58015653e+15,  1.10000000e+01, -5.00000000e+00])

Actually there's more reasons to love numpy arrays.

Arrays support element wise multiplication, addition, subtraction, and division by constants. Which is just what would happen to vectors, the mathematical object arrays were modeled after.

In [ ]:
2 * hoo_array + 1

Regular multiplication with arrays executes element wise multiplication

In [ ]:
hoo_array * hoo_array

Numpy arrays also handle matrix multiplication. What's that? Don't worry about it for now, just now that it exists.

In [ ]:
# define a new array, with easy to follow numbers
pl_array = np.array([1,2,3])

In [ ]:
#inner product, dot product
pl_array @ pl_array

In [ ]:
#matmul works as a dot product/ inner product as well
np.matmul(pl_array,pl_array)

In [ ]:
#use np.outer for outer product
first_matrix = np.outer(pl_array,pl_array)
first_matrix

### Array Creation

One of the great things about computers is that they can do tedious tasks for me, like make a 100 by 100 matrix.

In [ ]:
#feel the power
some_ones = np.ones((100,100))
some_ones

In [ ]:
#zeros works the same way, except with zeros
np.zeros(3)

Here are examples of other ways to make arrays. You can find even more [here](https://numpy.org/doc/stable/reference/routines.array-creation.html).

In [17]:
#creates an array of 21 values, evenly spaced from -10 to 10, inclusive
#notice how it always generates floats
np.linspace(-10,10,21)

array([-10.,  -9.,  -8.,  -7.,  -6.,  -5.,  -4.,  -3.,  -2.,  -1.,   0.,
         1.,   2.,   3.,   4.,   5.,   6.,   7.,   8.,   9.,  10.])

In [34]:
#also creates evely spaced values over a specificed range.
#similar to linspace, but less stable, and requires less inputs
#sometimes it generates floats, sometimes it generates ints
x_array = np.arange(10)
x_array

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [ ]:
#makes an array of 15 4s
np.repeat(4,15)

**Task7**:

Make an array from 87 to 104 (inclusive) with exactly 50 entries.

In [18]:
#ANSWER7

np.linspace(87, 104, 50)

array([ 87.        ,  87.34693878,  87.69387755,  88.04081633,
        88.3877551 ,  88.73469388,  89.08163265,  89.42857143,
        89.7755102 ,  90.12244898,  90.46938776,  90.81632653,
        91.16326531,  91.51020408,  91.85714286,  92.20408163,
        92.55102041,  92.89795918,  93.24489796,  93.59183673,
        93.93877551,  94.28571429,  94.63265306,  94.97959184,
        95.32653061,  95.67346939,  96.02040816,  96.36734694,
        96.71428571,  97.06122449,  97.40816327,  97.75510204,
        98.10204082,  98.44897959,  98.79591837,  99.14285714,
        99.48979592,  99.83673469, 100.18367347, 100.53061224,
       100.87755102, 101.2244898 , 101.57142857, 101.91836735,
       102.26530612, 102.6122449 , 102.95918367, 103.30612245,
       103.65306122, 104.        ])

Splicing works the same for arrays and lists.

**Task8**:

Grab the first through fifth entries of x_array.

In [37]:
#ANSWER8
listThing = x_array[0:5]
print(listThing)
#[*write your answer here*]

[0 1 2 3 4]


### Vectorized Operations

Numpy arrays come with many built in methods, many of which are vectorized operations. All you need to knnow about 'vectorized operations' is that: they are usually much faster than any other option, and how to use them.

There are two examples below. You can see others [here](https://numpy.org/doc/stable/reference/arrays.ndarray.html#array-methods).

In [ ]:
#finds the most positive (largest) value in the array
x_array.max()

In [ ]:
#finds the average of the values in x_array
x_array.mean()

## Pandas Dataframes

These are the data structure we will most often directly manipulate in this class. Paying special attention to this section will make your life easier later.

Like numpy, we need to import pandas, and we often use an abreviation.

In [ ]:
import pandas as pd

We can make a pandas dataframe from any numpy array.

In [ ]:
x_df = pd.DataFrame(x_array)
x_df

Notice how the one dimensional array was turned into a vertical column. Pandas dataframes frame (ahem) data around columns and column labels. There are row labels too, and these are refered to as indicies.

In [ ]:
some_ones_df = pd.DataFrame(some_ones)
some_ones_df

Pandas dataframes also have lots of methods, many of them vectorized operations, like numpy. This is because pandas was built on top of numpy.

In [ ]:
#finds the largest value in column 0 of x_df
x_df.max()

In [ ]:
#sums each row
some_ones_df.sum(axis=0)

### Slicing Dataframes

For the next part, we're going to need a more interesting example than some_ones_df.

To get a more interesting dataframe, I'm going to pull one from my github using pd.read_csv. This is the most common way you'll get data in this course, but pandas offers more versitile options. You can view them [here](https://pandas.pydata.org/docs/reference/io.html).

In [39]:
import pandas as pd

#make a string of the webpage where my data lives
url_basketball = 'https://raw.githubusercontent.com/thedarredondo/data-science-fundamentals/main/Data/basketball2324.csv'
#feed that string into the pandas read csv function
basketball = pd.read_csv(url_basketball)

In [40]:
basketball

Player   Pos  Age   Tm   G  GS    MP   FG   FGA    FG%  \
0            Precious Achiuwa  PF-C   24  TOT  74  18  1624  235   469  0.501   
1                 Bam Adebayo     C   26  MIA  71  71  2416  530  1017  0.521   
2                Ochai Agbaji    SG   23  TOT  78  28  1641  178   433  0.411   
3                Santi Aldama    PF   23  MEM  61  35  1618  247   568  0.435   
4    Nickeil Alexander-Walker    SG   25  MIN  82  20  1921  236   538  0.439   
..                        ...   ...  ...  ...  ..  ..   ...  ...   ...    ...   
567            Thaddeus Young    PF   35  TOT  33   6   439   65   108  0.602   
568                Trae Young    PG   25  ATL  54  54  1942  433  1008  0.430   
569            Omer Yurtseven     C   25  UTA  48  12   545   99   184  0.538   
570               Cody Zeller     C   31  NOP  43   0   320   26    62  0.419   
571               Ivica Zubac     C   26  LAC  68  68  1794  337   519  0.649   

     ...  ORB  DRB  TRB  AST  STL  BLK  TOV   PF   PTS  Player-additional  
0    ...  191  296  487   97   46   68   83  143   565          achiupr01  
1    ...  159  578  737  278   81   66  162  159  1367          adebaba01  
2    ...   74  142  216   83   47   44   64  117   455          agbajoc01  
3    ...   72  280  352  138   43   54   69   89   654          aldamsa01  
4    ...   35  132  167  204   64   42   76  143   655          alexani01  
..   ...  ...  ...  ...  ...  ...  ...  ...  ...   ...                ...  
567  ...   46   57  103   57   23    5   15   49   137          youngth01  
568  ...   23  126  149  583   72   11  235  109  1389          youngtr01  
569  ...   72  136  208   29    8   18   37   52   222          yurtsom01  
570  ...   48   64  112   39    9    5   16   45    76          zelleco01  
571  ...  196  430  626   93   22   83   79  180   794          zubaciv01  

[572 rows x 30 columns]

Its easy to call the column names and rows. You can also call the shape, but note that printing the dataframe also tells you the shape.

In [ ]:
basketball.columns

In [ ]:
basketball.index

In [ ]:
basketball.shape

We call columns by name.

In [ ]:
basketball.Player

In [ ]:
basketball['Player']

In [ ]:
#the double brackets returns a data frame
basketball[['Player']]

The second brackets method is required when we want a column with funky characters in its name.

Want to slice multiple columns?

In [ ]:
basketball[['Player','FG%']]

Want to slice multiple rows and multiple columns?

In [ ]:
#use loc to grab the zeroth, second, and 207th row.
basketball[['Player', 'FG%']].loc[[0,4,207]]

**Task9**:

Find the names of the Players, feild goal attempts (FGA) and total points (PTS) of the players in the 87th and  94th row.

In [41]:
#ANSWER9

basketball[['Player', 'FGA', 'PTS']].loc[[87,94]]

Player  FGA  PTS
87  Wendell Carter Jr.  425  603
94        Max Christie  241  284

### Query and Bools

Because I provide clean, moderate to small sized data sets in this course, most of our data wrangling problems can be solved with query.

Query makes use a of ab built in data type we haven't mentioned yet, called a bool.

I'll give you an example of query, and then I'll define bools.

In [ ]:
#finds all players with more offensive rebounds than defensive rebounds
#and who have also played more than 100 minutes
hustle_players = basketball.query('ORB >= DRB and MP >= 100')
hustle_players

Bools are True or False. True is 1, and False is 0. The key words True and False cannot be variables

In [ ]:
True = 2

They come into play when I apply boolean operators. For example, if I want to know if 5 equals 5.

In [ ]:
#does 5 equal 5?
5 == 5

In [ ]:
#does 5 equal 'five'?
5 == 'five'

True and False are also the numbers 1 and 0 respectively. This can come in handy

In [ ]:
#True equals 1
True == 1

In [ ]:
#True not equaling 1 is False
True != 1

In [ ]:
#Adding bools counts the number of trues
False + True + True + False + False

Here's an example each of and, or, and not.

In [ ]:
#both sides must be true for and
(True == 1) and (False != 0)

In [ ]:
#only one side (or more) needs to be true for or
(True == 1) or (False != 0)

In [ ]:
#not swaps True for False, and False for True
not 0, not False

Here are the other four boolean operations.

In [ ]:
#strictly less than, and strickly greater than
0 < 0, 4 > 5

In [ ]:
#less than or equal to, and greater than or equal to
5 <= 5, 6 >= 5

Now that you know what bools are, use my example query from above to complete the task below.

**Task10**:

Find all the players that have played more than 41 games (G), and have more turn overs (TOV) than assists (AST).

In [49]:
#ANSWER10

playerNames = []

for index, i in basketball.iterrows():
  if(i['G'] > 41 and i['TOV'] > i['AST']):
    playerNames.append(i['Player'])

print(playerNames)

['Deandre Ayton', 'Marvin Bagley III', 'Bojan Bogdanović', 'Bol Bol', 'Chris Boucher', 'John Collins', 'Andre Drummond', 'Rudy Gobert', 'Jaxson Hayes', 'GG Jackson II', 'Jaren Jackson Jr.', 'Walker Kessler', 'JaVale McGee', 'Chimezie Metu', 'Naz Reid', 'Nick Richards', 'Luka Šamanić', 'Myles Turner', 'Moritz Wagner', 'Cam Whitmore', 'James Wiseman', 'Christian Wood', 'Omer Yurtseven']


## For loops

Most of the time, there is some [dataframe method](https://pandas.pydata.org/docs/reference/frame.html) that does whatever we want, in terms of data wrangling.

Occasionally, there will be no built in method that quite does what we need.

In those cases, we will use a for loop. For loops often involve if statements, so we'll introduce those as well.

I'll start with a simple example that isn't all that useful.

In [ ]:
#initialize a variable to update during the for loop
x = 0

#initialize an iterable to run our for loop over
onetoten = np.arange(1,11)

#the actual for loop
for i in onetoten:
    x = x + i

In [ ]:
#the sume of the numbers from one to ten
x

Most of the time, we'll avoid for loops for anything but plotting, because their are often better vectorized solutions available.

In [ ]:
#could have just used the vectroized solution
onetoten.sum()

Here's a more relevant example:

While playing with the 2023/2024 NBA basketball data, I wanted to pull up Lebron James' stats. And I figured I might as well pull all the other players with James in their name, since there couldn't be that many.

I searched for a while, and couldn't find a nice vectorized solution that uses the pandas methods.

So I wrote this for loop. It has too new things from the previous loop:

- a function, enumerate, which allows me to grab the item in my iterable and the index of that item.
- an if statement. They work like this: if the condition in them is met, execute the indented code. If not, then continue with the rest of the for loop.

Most for loops we need in this class will be at least this complicated. Usually more so.

In [ ]:
#create an array of all zeros the same length as one of the columns in basketball
james_index = np.zeros_like(basketball.Player)

#index is the number of the row, name is the string that represents player name
#enumerate is a function that breaks basketball.Player into index and name
for index, name in enumerate(basketball.Player):

    #if the string 'James' is in the string of player name
    if 'James' in name:

        #then put a True value, or 1, in the corresponding James index tracker
        james_index[index] = 1

#adds up the array elements of james_index
#this helps me know if my loop went awry or not
#but otherwise has nothing to do with the for loop.
james_index.sum()

In [ ]:
#always a good idea to print out any data set you've created
#even if you're expecting it to be uninteresting
james_index

In [ ]:
#this grabs all the rows/players with James in their name
basketball.loc[james_index == True]

This next task will require all the most relevant data manipulation skills we've learned thus far. Try to copy and paste code from this document, and then modify as needed. Remember, your code doesn't need to work the first, or even tenth, time you run it.

**Task11**:

Display all players--and their stats--who are PG (point guards), have a 0.9 FT% (free throw percentage) or higher  AND more than 40 FTA (free throw attempts) for the 2023/2024 NBA season.

In [48]:
#ANSWER11

playerInfo = []

for index, i in basketball.iterrows():
  if(i['Pos'] == 'PG' and i['FT%'] >= 0.9 and i['FTA'] > 40):
    playerInfo.append(i['Player'])

print(playerInfo)

['Mike Conley', 'Stephen Curry', 'Aaron Holiday', 'Damian Lillard']


In [ ]:
#ANSWER11 cont.

#[*write your answer here*]

## Exporting Data from Colab

I don't plan on making you do this as part of the course, but this is still a skill I want you to have.

As usual, I'm only covering a tiny bit of what's available. For more info, see: [Importing/Exporting files using pandas](https://pandas.pydata.org/docs/reference/io.html)

Let's say I have a colleague for who doesn't know how to code and will not ever learn how, but they still need the data I've manipulated/generated in order to do their job.

In this case, the path of least resistance is to export the data from Colab to my local machine, and then put it in a shared cloud folder (like google drive) and/or send it to them in an email.

Let's practice this with the James data, shown below.

In [ ]:
james_df = basketball.loc[james_index == True]
james_df

Here's how I can turn this into a csv, a format most programs (like excel) can read:

In [ ]:
#creates a csv file on whatver machine the code is run on.
#I need to name the file as well.
james_df.to_csv('NBA_Jameses.csv')

Next steps:

1. Look at the side panel--to the left of all the code and text blocks.

2. Click on the icon that looks like a tiny manilla folder.

3. Find NBA_Jameses.csv, hover your mouse over it, and click on the three stacked dots to the right of it.

4. Click download.

From there, its the same as downloading any file from the internet onto your computer.

While you're there, notice that you can upload files to the google server where all your code is being run, and even mount your google drive account.

## Further Reading

As promised, there is much I didn't cover. Here's lip service to some of that left out information.



### Tensors and xarrays

Notable datastructures I didn't cover, but that are relevant for modern machince learning: tensors, and xarrays.

Tensors are multi dimensional arrays; technically numpy can make them, but tensorflow and Pytorch are two libraries that implement tensors more optimized for statistical learning.

xarray is its own python library, and it's basically what would happen if a dataframe and a dict had the perfect child.

We use tensors are for doing math with high dimensional data, and we use xarrays to keeping track of a lot of seperate, but related, datasets.

We will hardly use tensors in this class, but we'll use xarrays all the time. We'll discuss them more as they come up.


### Data Wrangling

Data Wrangling is what happens when the data you want is not a csv . That's a bit of a joke and an exaggeration . . . but its also not.

When I say csv, I mean nice, neat rows and columns, with everything lined up and seperated by commas.

Here are various ways real world data can make us suffer:
- it's not collected in rows and columns
- it's collected in rows and columns, but so disorganized that the rows and columns don't actually mean anything.
- it's collected in neat rows and columns, but the the entries into those rows and columns are inconsistent in format.
- everything is in rows and columns, neat and consistent throughout--but our functions are still reading the data wrong (this happened to me with a csv file that wasn't actually seperated by commas, but by semi colons instead).
- like the last point, except everything was read correctly, but now there are missing values.
- any combination of the above.

Even if everything is actually perfect--which does happen--data sets can be so large that doing anything with them at all is a chore.

For class notes and exercises, I've limited my examples to perfectly clean, small to medium sized data.

Projects are open ended. On projects, it's up to you if you want to choose a pre-cleaned data set or not.

## Summary

We covered quite a bit in our first unit.

We:

- learned the two main ways python deals with numbers, ints and floats
- covered strings, lists, and dicts, which we can use to store multiple pieces of data linked together.
- introduced the numpy array, the most common way to mathematically operate on collections of data of the same type in python
- manipulated real world tabular data using pandas
- learned how to import and export data from pandas.

I expect most of y'all to feel overwhelmed and confused, especially if this is your first coding class. That's good! It means you're thinking about this material.

Remember, I'm not asking or expecting you to become an expert data scientist. Our goal is to get vulnearble and expose ourselves to how data science/modern statistics is done. Feel good if you've done that!